### ISSUE TITLE
How many request can it handle concurrently ?

### ISSUE THREAD

**Issue by somprabhsharma** [Microsoft/BotBuilder#1250]
I have a 10000 users chatting with the bot simultaneously, then will Microsoft Bot Framework be able to handle this many requests concurrently? How you handle requests ?

**Comment by willportnoy** [comment]
Activity HTTP requests are handled in parallel.  Of course, you may find the channel itself might throttle your messages, but we don't add throttling on top of that.

**Comment by somprabhsharma** [comment]
Is there any plan for you to open source the part of code you have at your end which interact with channels directly ? Or is there any way to get that code so that we can deploy it on a our own server that can handle requests for 15 million people to various channels and can make things faster by using various mechanism and make sure that the messages are getting delivered, if not we will resend the messages.?

**Comment by dandriscoll** [comment]
You can certainly implement retries on your end, if you'd like. Our servers are geo-distributed and are heavily parallelized, so if you send many requests at once, they will be sent to a datacenter nearest you and will likely be handled by different machines.

**Comment by somprabhsharma** [comment]
Yeah, you are right. But can you please explain how to implement retries at my end ? when I am not getting any delivery confirmation of the message. How can I know whether the message got delivered or not so that I can resend it.

**Comment by dandriscoll** [comment]
Most of our servers are coded so that if you receive a 2xx HTTP status code, you do not need to retry. We have accepted the message and will make sure it gets delivered.

There are a few cases where a message is dropped if it isn't properly formatted, although these are bugs and we are fixing them.

Clearly, a connection failure is something you have to retry. A 4xx and 5xx code *may* be a reason to retry, although that may indicate an error in your data payload.

Does that help clarify? Let me know if you have additional questions.

**Comment by somprabhsharma** [comment]
Yes that helps. You can close the issue for now. Thank you.

**Comment by willportnoy** [comment]
Thank you for the discussion - feel free to open a new issue with more questions.
